framework/src/wifi_manager/wifi_manager.c

```diff
@@ -0,0 +1,440 @@
+		ndbg("wifi_utils_deinit fail");
+		return WIFI_MANAGER_FAIL;
+	}
+
+	return WIFI_MANAGER_SUCCESS;
+}
+
+wifi_manager_result_e wifi_manager_set_mode(wifi_manager_mode_e mode)
+{
+	wifi_utils_result_e result = WIFI_UTILS_SUCCESS;
+
+	if (mode != STA_MODE && mode != SOFT_AP_MODE) {
+		return WIFI_MANAGER_INVALID_ARGS;
+	}
+	// STA mode -> SOFT AP mode
+	if (g_manager_info.mode == STA_MODE && mode == SOFT_AP_MODE) {
+		wifi_utils_softap_config_s soft_ap_config;
+
+		soft_ap_config.channel = g_wifi_softap_config->channel;
+		soft_ap_config.ap_crypto_type = WIFI_UTILS_CRYPTO_AES;
+		soft_ap_config.ap_auth_type = WIFI_UTILS_AUTH_WPA2_PSK;
+		strncpy(soft_ap_config.ssid, g_wifi_softap_config->ssid, sizeof(g_wifi_softap_config->ssid));
+		soft_ap_config.ssid_length = strlen(g_wifi_softap_config->ssid);
+		strncpy(soft_ap_config.passphrase, g_wifi_softap_config->passphrase, sizeof(g_wifi_softap_config->passphrase));
+		soft_ap_config.passphrase_length = strlen(g_wifi_softap_config->passphrase);
+		soft_ap_config.inform_new_sta_join = g_manager_info.wmcb->softap_sta_join;
+
+		wifi_mutex_acquire(&w_mutex, WIFI_UTILS_FOREVER);
+
+		g_new_sta_join = soft_ap_config.inform_new_sta_join;
+		result = wifi_utils_start_soft_ap(&soft_ap_config);
+
+		if (result != WIFI_UTILS_SUCCESS) {
+			ndbg("Start soft ap mode fail");
+			wifi_mutex_release(&w_mutex);
+			return WIFI_MANAGER_FAIL;
+		}
+
+		if (start_dhcp_server() != WIFI_UTILS_SUCCESS) {
+			ndbg("start DHCP server Failed\n");
+			return WIFI_MANAGER_FAIL;
+		}
+
+		g_manager_info.mode = SOFT_AP_MODE;
+		g_manager_info.status = CLIENT_DISCONNECTED;
+		wifi_mutex_release(&w_mutex);
+
+		nvdbg("Wifi Change STA MODE -> SOFT AP MODE");
+	}
+	// SOFT AP mode -> STA mode
+	else if (g_manager_info.mode == SOFT_AP_MODE && mode == STA_MODE) {
+		wifi_mutex_acquire(&w_mutex, WIFI_UTILS_FOREVER);
+
+		result = wifi_utils_stop();
+		if (result != WIFI_UTILS_SUCCESS) {
+			ndbg("Wifi stop fail");
+			wifi_mutex_release(&w_mutex);
+			return WIFI_MANAGER_FAIL;
+		}
+
+		stop_dhcp_server();
+
+		result = wifi_utils_start_sta();
+		if (result != WIFI_UTILS_SUCCESS) {
+			ndbg("start STA fail (change STA mode fail)");
+			wifi_mutex_release(&w_mutex);
+			return WIFI_MANAGER_FAIL;
+		}
+
+		g_manager_info.mode = STA_MODE;
+		g_manager_info.status = AP_DISCONNECTED;
+		wifi_mutex_release(&w_mutex);
+
+		nvdbg("Wifi Chnage SOFT AP MODE -> STA MODE");
+	}
+
+	return WIFI_MANAGER_SUCCESS;
+}
+
+wifi_manager_result_e wifi_manager_get_info(wifi_manager_info_s *info)
+{
+	wifi_mutex_acquire(&w_mutex, WIFI_UTILS_FOREVER);
+	*info = g_manager_info;
+	wifi_mutex_release(&w_mutex);
+
+	return WIFI_MANAGER_SUCCESS;
+}
+
+wifi_manager_result_e wifi_manager_set_softap_config(wifi_manager_softap_config_s *config)
+{
+	if (config == NULL) {
+		return WIFI_MANAGER_FAIL;
+	}
+
+	if (!g_wifi_softap_config) {
+		g_wifi_softap_config = (wifi_manager_softap_config_s *)malloc(sizeof(wifi_manager_softap_config_s));
```

Comment:
I can't find codes to free g_wifi_softap_config. Only allocation is here. Is it intended?

Reply:
Very good comment!
Actually this api is not required if softap configuration information is delivered when wifi_manager_set_mode() is called.